spec/bucketing_spec.rb

```diff
@@ -108,47 +108,82 @@ def get_bucketing_id(user_id, entity_id=nil)
-    expected_bucketing_id = get_bucketing_id('test_user', '101')
+    expected_bucketing_id = get_bucketing_key('ppid8','101')
     expect(bucketer).to receive(:generate_bucket_value).once.with(expected_bucketing_id).and_call_original
+    
     experiment = config.get_experiment_from_key('group1_exp1')
-    bucketer.bucket(experiment, 'test_user')
+    bucketer.bucket(experiment,'ppid8','test_user')
+
+    expect(spy_logger).to have_received(:log)
+    .with(Logger::INFO, "User 'test_user' is not in experiment '#{experiment['key']}' of group #{experiment['groupId']}.")
   end
 
   it 'should return nil when user is in an empty traffic allocation range due to sticky bucketing' do
     expect(bucketer).to receive(:find_bucket).once.and_return('')
     experiment = config.get_experiment_from_key('test_experiment')
-    expect(bucketer.bucket(experiment, 'test_user')).to be_nil
+    expect(bucketer.bucket(experiment,'bucket_id_ignored', 'test_user')).to be_nil
     expect(spy_logger).to have_received(:log)
                       .with(Logger::INFO, "User 'test_user' is in no variation.")
     expect(spy_logger).to have_received(:log)
                       .with(Logger::DEBUG, "Bucketed into an empty traffic range. Returning nil.")
   end
 
+  describe 'Bucketing with Bucketing Id' do
+    # Bucketing with bucketing ID
+    it 'should bucket to a variation different than the one expected with the userId' do
+      experiment = config.get_experiment_from_key('test_experiment')
+
+      # Bucketing with user id as bucketing id - 'test_user111127' produces bucket value < 5000
+      expected_variation = config.get_variation_from_id('test_experiment','111128')
+      expect(bucketer.bucket(experiment,'test_user', 'test_user')).to be(expected_variation)
+
+      # Bucketing with bucketing id - 'any_string789111127' produces bucket value btw 5000 to 10,000
+      expected_variation = config.get_variation_from_id('test_experiment','111129')
+      expect(bucketer.bucket(experiment,'any_string789', 'test_user')).to be(expected_variation)
+    end
+
+    # Bucketing with invalid experiment key and bucketing ID
+    it 'should return nil with invalid experiment and bucketing ID' do
+      expect(bucketer.bucket(config.get_experiment_from_key('invalid_experiment'),'some_id', 'test_user')).to be(nil)
+    end
+
+    # Bucketing with grouped experiments and bucketing ID
+    it 'should bucket expectedly when bucketing id in grouped experiments'  do
+      experiment = config.get_experiment_from_key('group1_exp1')
+
+      expected_variation = config.get_variation_from_id('group1_exp1','130001')
+      expect(bucketer.bucket(experiment,'pid', 'test_user')).to be(expected_variation)
+
+      expected_variation = config.get_variation_from_id('group1_exp1','130002')
+      expect(bucketer.bucket(experiment,'123456789', 'test_user')).to be(expected_variation)
+    end
+  end
+
```

Comment:
Please add the cases when the bucketing key fallbacks to the user ID. This happens when the bucketing key is null. So, to clarify:
• Bucketing with bucketing ID
-Non null bucketing ID
-Null bucketing ID
• Bucketing with invalid experiment key and bucketing ID
-Non null bucketing ID
-Null bucketing ID
• Bucketing with grouped experiments and bucketing ID
-Non null bucketing ID
-Null bucketing ID

I've updated the doc with these corner cases. 

You can look at the PHP example: 
https://github.com/optimizely/php-sdk/blob/master/tests/BucketerTest.php#L298

Reply:
@alda-optimizely  I have a question regarding this. Please see my comments here:
https://github.com/msohailhussain/php-sdk/pull/6

Reply:
Yea you're right. I don't remember why I did that...maybe it was some legacy implementation that I forgot to update? The main goal for the test is to make sure that the bucketing ID is being used and not the userID for the bucketing. So, userID should bucket to the non expected variation. I've updated my PHP tests.